### PR TITLE
Add a CLI progress bar

### PR DESCRIPTION
**Pull request by collijk**
### Description

*Category*: feature
*Issues*: N/A

The main thrust of this PR is to add a progress bar to the CLI and to clean up the interactive simulation progress bar implementation.

Example CLI usage:
image

Example interactive sim usage from REPL: 
image

Example interactive sim usage from notebook:
image

**Detailed changes**
- Add `tqdm` to framework dependencies
- Add a `steps_remaining` property to simulation clocks so they can tell us the remaining number of simulation steps.
- Add a private context manager to the `InteractiveContext` to abstract away the behavior of adjusting the simulation step size.
- Add a `take_steps` method to the base `SimulationContext` that is smart about the balance between verbose logging and using a progress bar.
- Make the time stepping methods on the base `SimulationContext` and the `InteractiveContext` more consistent and clean up their implementations a bit.
- Replace custom progress bar code in the `InteractiveContext` with `tqdm`'s implementation so we don't have to think about it when it breaks (as it's already broken).

### Testing
CI + manual testing

### CODE REVIEW

**Review by stevebachmeier — APPROVED** [review]
nice

File: src/vivarium/framework/engine.py [hunk 1]
```diff
@@ -202,16 +203,28 @@ def initialize_simulants(self):
...
             self.time_step_emitters[event](self._population.get_population(True).index)
+
         self._clock.step_forward()
 
-    def run(self):
-        while self._clock.time < self._clock.stop_time:
-            self.step()
+    def take_steps(self, number_of_steps: int, progress_bar: bool = False) -> None:
+        log_step = not progress_bar
+        with tqdm.tqdm(total=number_of_steps, disable=not progress_bar) as pbar:
```
Comment (rmudambi) [inline src/vivarium/framework/engine.py:242]: Is it not possible for this logic to live inside the logging system? It feels more natural there.

File: src/vivarium/interface/interactive.py [hunk 2]
```diff
@@ -234,6 +259,18 @@ def get_component(self, name: str) -> Any:
         """
         return self._component_manager.get_component(name)
 
+    @contextmanager
+    def _adjust_step_size(self, step_size: Timedelta = None):
+        old_step_size = self._clock.step_size
```
Comment (rmudambi) [inline src/vivarium/interface/interactive.py:264]: Can you add a docstring?

File: src/vivarium/interface/interactive.py [hunk 3]
```diff
@@ -118,39 +131,51 @@ def run_until(self, end_time: Time, with_logging: bool = True) -> int:
...
-
-    def take_steps(
-        self, number_of_steps: int = 1, step_size: Timedelta = None, with_logging: bool = True
-    ):
-        """Run the simulation for the given number of steps.
+            msg = f"Provided time must be an instance of {type(self._clock.time)}"
+            raise TypeError(msg)
+
+        step_size = step_size if step_size is not None else self._clock.step_size
+        number_of_steps = int(ceil((end_time - self._clock.time) / self._clock.step_size))
```
Comment (rmudambi) [inline src/vivarium/interface/interactive.py:138]: Would it make sense to make `clock.steps_remaining` a function that takes an optional parameter of end time? This is the same logic that you are using there.

File: src/vivarium/framework/engine.py (hunk 1 above)
Comment (collijk) [reply src/vivarium/framework/engine.py:242]: I think it is very not advisable.  The logger is a very thin wrapper around normal logging conventions.  Making good and informative progress bars requires a lot of context-specific state information that I think would be hard to generalize reasonably into the logging system.

On a more theoretical point, I think of progress bars as kind of an animation trick for your terminal that obscures the guarantees of a logging stream (a stream of discrete messages), which make me a bit hesitant to conflate the two.

File: src/vivarium/interface/cli.py [hunk 4]
```diff
@@ -93,6 +93,12 @@ def simulate():
     is_flag=True,
     help="Suppresses all logging except for warnings and errors.",
 )
+@click.option(
+    "--progress-bar",
+    "--pb",
```
[the remark this answers is not in the document]
Comment (collijk) [reply src/vivarium/interface/cli.py:98]: We used both in all the covid libraries without much issue.

File: src/vivarium/interface/interactive.py (hunk 3 above)
Comment (collijk) [reply src/vivarium/interface/interactive.py:138]: I don't think so.  I think of the clock as a closed system with a current time and step size and some functions for advancing time, and not really a calculator for time-based inquiries.  But I'm not adamant about it.

Comment (rmudambi) [reply src/vivarium/interface/interactive.py:138]: As a compromise, you could have a utility function in the time module that is called in both locations?

**Review by rmudambi — APPROVED** [review]

**Pull request closed without merging**